### PR TITLE
VIMS-472 Add "includeEmptyLots" option to endpoints which return Stoc…

### PR DESCRIPTION
**Pull request by benleibert**
…kCard(s).

### CODE REVIEW

File: src/main/java/org/openlmis/stockmanagement/controller/StockCardController.java [hunk 1]
```diff
@@ -342,7 +346,26 @@ public ResponseEntity processStock(@PathVariable long facilityId,
...
+    //Convenience method, calling TruncateStockCardEntries() and removeEmptyLotsFromStockCard()
+    private void filterEntries(StockCard stockCard, Integer entryCount, boolean includeEmptyLots)
+    {
+        TruncateStockCardEntries(stockCard, entryCount);
+        if(!includeEmptyLots)
+            removeEmptyLotsFromStockCard(stockCard);
+    }
+
+    //Filter stockCard.entries such that it contains only the first entryCount number of items
+    private void TruncateStockCardEntries(StockCard stockCard, Integer entryCount) {
```
Comment (chunky56) [inline src/main/java/org/openlmis/stockmanagement/controller/StockCardController.java]: May want to lowercase first letter of method to match Java conventions and the other code.

File: src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java [hunk 2]
```diff
@@ -131,17 +130,32 @@ public void setupEvent() {
...
+    lot.setManufacturerName("Manufacturer_of_" + id);
     lot.setManufactureDate(new Date());
     lot.setExpirationDate(new Date());
-    event.setLot(lot);
+    //event.setLot(lot);
+    return lot;
+  }
+
+  //Associate two lots with the specified StockCard - one lot for which we have stockOnHand, and one for which we don't
+  public void addLotsToStockCard(StockCard card)
```
Comment (chunky56) [inline src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java]: Method name is not clear that it simply sets the stock card to have two lots; expecting that lots are added each time the method is called.

File: src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java [hunk 3]
```diff
@@ -172,14 +186,79 @@ public void shouldSucceedWithValidAdjustment() {
...
     StockCardEntry entry = new StockCardEntry(dummyCard, StockCardEntryType.ADJUSTMENT, event.getQuantity() * -1);
     entry.setAdjustmentReason(reason);
-    verify(service).addStockCardEntries(Collections.singletonList(entry));
+    verify(stockCardService).addStockCardEntries(Collections.singletonList(entry));
     assertThat(response.getStatusCode(), is(HttpStatus.OK));
   }
+
+
+  @Test
+  public void shouldOnlyReturnEmptyLotsWhenRequested()
```
Comment (chunky56) [inline src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java:201]: Probably want to separate this into multiple test cases.

File: src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java [hunk 4]
```diff
@@ -172,14 +186,79 @@ public void shouldSucceedWithValidAdjustment() {
...
+    includeEmptyLots = true;
+
+
+    addLotsToStockCard(dummyCard);
+    response = controller.getStockCard(facilityId, productId, numEntries, includeEmptyLots);
+    stockCard = (StockCard)response.getBody();
+    assertEquals( 2, stockCard.getLotsOnHand().size());
+
+    addLotsToStockCard(dummyCard);
+    controller.getStockCardById(facilityId, stockCardId, numEntries, includeEmptyLots);
```
Comment (chunky56) [inline src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java]: Not assigning the return value of this call to response? Then code on next line will just get value of previous call.

File: src/main/java/org/openlmis/stockmanagement/controller/StockCardController.java (hunk 1 above)
Comment (benleibert) [inline src/main/java/org/openlmis/stockmanagement/controller/StockCardController.java]: Good catch - thank you. I fixed that mistake.

File: src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java (hunk 4 above)
Comment (benleibert) [inline src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java]: Good catch - thank you! I fixed that mistake.

File: src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java (hunk 2 above)
Comment (benleibert) [inline src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java]: I renamed the method to "associateTestLotsWithStockCard."

File: src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java (hunk 3 above)
Comment (benleibert) [inline src/test/java/org/openlmis/stockmanagement/controller/StockCardControllerTest.java:201]: I had considered doing so, and wonder if this is a matter of preference. I don't think this is what "Effective Unit Testing" would call a "Split Personality" test, though. Would certainly be interested in discussing.

**Pull request merged**